src/StoreMeta.h

```diff
@@ -127,7 +127,9 @@ class StoreMeta
 
     virtual char getType() const = 0;
     virtual bool validLength(int) const;
-    virtual bool checkConsistency(StoreEntry *) const;
+    /// Checks that this is a valid StoreMeta.
+    /// Kids may store value into StoreEntry.
+    virtual void applyTo(StoreEntry *) const;
```

Comment:
I see two problems with this solution:

A) The new API still suffers the old checkConsistency() problem of having an inconsistent/invalid object. We just changed the "consistency" in the method name (and return type) to "valid" in the method description (and exceptions), but the essence of that problem remains -- Squid has code that operates on apparently invalid/inconsistent C++ objects. Those objects should not exist in the first place. Whether we are creating objects we should not be creating or whether we are applying consistency checks to the wrong object(s) is an open question.

B) If I am reading the code correctly, there are actually at least two ways to "apply" a StoreMeta to StoreEntry. We should not suggest a single applyTo() method for expressing multiple application algorithms. To find those different algorithms, look for `StoreMetaUnpacker aBuilder` occurrences:

1. Rock::HeaderUpdater::parseReadBytes() applies nothing -- individual StoreMeta entries are ignored AFAICT. I suspect we can ignore them because we already loaded/handled them before and are now just trying to find the beginning of HTTP headers that follow them, but I am not sure. It may be OK to ignore this use case because it does not deal with individual StoreMeta entries, but we should keep it in mind...
2. store_client::unpackHeader(). This is the checkConsistency() code you have improved. This is the only code using this virtual method. It updates the existing StoreEntry object based on disk- or shared memory cache-stored metadata.
3. storeRebuildParseEntry(). This code uses a visitor pattern to implement a different flavor of applyTo(). Instead of being located inside a StoreMeta method, this application algorithm lives in a switch statement inside InitStoreEntry (i.e. the "visitor")!

Since different application code runs inside two different functions, one of which is not a StoreMeta method at all, we cannot say that StoreMeta encapsulates how to update StoreEntry based on cached metainformation. We could split applyTo() into two methods, moving InitStoreEntry switch code into the second one -- applyToBeingLoadedEntry()) and applyToBeingIndexedEntry() -- but it feels a bit wrong to ask a rather primitive StoreMeta object to update a StoreEntry, especially since some future updates might require multiple StoreMeta objects or might delay all updates to do an overall consistency check first. I think such updates should be done by higher-level code that knows the update context/needs. In other words, InitStoreEntry code got it right (in principle) and store_client::unpackHeader() should follow that lead instead.

The existence of a big `switch` statement inside StoreMeta::checkConsistency() supports that idea -- the checkConsistency() code already leans towards centralizing rather than virtualizing each StoreMeta application because folks do not want to create a lot of virtual methods that do almost nothing. They want to group identical StoreMeta checking/application algorithms together.

Overall, there is a lot of virtualization surrounding StoreMeta, but, in the end, we still have at least three StoreMeta type-based `switch` statements and `void*` data storage for each StoreMeta class (with associated type casts).

In all cases, StoreMeta objects do not survive/exist beyond the single StoreEntry object application. We are dynamically allocating a lot of temporary polymorphic objects and often immediately downgrading back to hard-coded switch statements to handle (groups of) type-specific validation and/or application.

Instead of renaming checkConsistency(), I think we should refactor the whole StoreMeta mess. I will make a more specific suggestion later, but if you see any serious flaws in the above (rather rough!) critique of the existing situation, please let me know. No code changes are requested at this time.

Reply:
I agree that the current StoreMeta-related implementation with both class hierarchy and big `switches` looks like overengeneering. I think that a better design would use one of these two approaches.

Reply:
@eduard-bagdasaryan, I have removed the StoreMeta hierarchy in #156. Please review that internal PR. If you agree with the overall direction of that change and think that upcoming Vary ID support changes will work with this new approach, then please fix/adjust anything I have missed and test both Rock and UFS swap out/in operations to make sure the changes are compatible with old cache_dirs.

Reply:
> upcoming Vary ID support changes will work with this new approach

I see no obstacles in applying the new approach to PR150:  we would need to move the IDs parsing from StoreMetaVaryId to a function instead, e.g., Store::GetNewSwapMetaVaryId()) and adjust the big switch in store_client::unpackHeader(), similar to the existing STORE_META_VARY_HEADERS handling.

Reply:
>  test both Rock and UFS swap out/in operations to make sure the changes are compatible with old cache_dirs

Done, no problems observed.